### PR TITLE
Publish docker images for PRs from forks to GCR.

### PR DESCRIPTION
**Pull request by isarkis**
GitHub Actions doesn't allow publishing images to GHR from pull requests created from forks.  As a workaround, for linux workloads, we will run docker image builds on self-hosted runners, publish images to GCR, and run linux workloads in those images.

### CODE REVIEW

**Comment by isarkis** [comment]
This PR can only be pushed after we build new linux runner (see cl/498104088 and http://go/lbreview/239040)

File: .github/actions/docker/action.yaml [hunk 1]
```diff
@@ -39,11 +40,23 @@ runs:
     # were changed we need to use an existing image (e.g. main).  In cases where docker image is rebuilt we have
     # to use tag generated by the image build.
     - name: Set Docker Tag
-      if: ${{ (steps.changed-files.outputs.any_changed == 'false') && (github.event_name == 'pull_request') }}
+      id: set-docker-tag-presubmit-non-fork
+      if: ${{ (steps.changed-files.outputs.any_changed == 'false') && (github.event_name == 'pull_request') && (github.repository == github.event.pull_request.head.repo.full_name) }}
```
Comment (andrewsavage1) [inline .github/actions/docker/action.yaml:44]: I believe you can just use the boolean `github.event.pull_request.head.repo.fork`

File: .github/actions/docker/action.yaml [hunk 2]
```diff
@@ -39,11 +40,23 @@ runs:
...
       run: echo "DOCKER_TAG=ghcr.io/${{github.repository}}/${{inputs.docker_image}}:${GITHUB_BASE_REF%.1+}" >> $GITHUB_ENV
       shell: bash
+    - name: Set Docker Tag
+      id: set-docker-tag-presubmit-fork
+      if: ${{ (steps.changed-files.outputs.any_changed == 'true') && (github.event_name == 'pull_request') && (github.repository != github.event.pull_request.head.repo.full_name) }}
```
Comment (andrewsavage1) [inline .github/actions/docker/action.yaml:49]: Same as above

File: .github/actions/docker_win/action.yaml [hunk 3]
```diff
@@ -62,11 +62,11 @@ runs:
         composeArguments: "--no-start"
         composeFile: "docker-compose-windows.yml"
     - name: Tag images
-      if: env.need_to_build == 'true'
+      if: ${{ (env.need_to_build == 'true') && ((github.event_name != 'pull_request') || (github.repository == github.event.pull_request.head.repo.full_name)) }}
```
Comment (andrewsavage1) [inline .github/actions/docker_win/action.yaml:65]: Same as above

File: .github/actions/docker_win/action.yaml [hunk 4]
```diff
@@ -62,11 +62,11 @@ runs:
...
       run: |
         docker tag cobalt-${{inputs.service}} $DOCKER_TAG
       shell: bash
     - name: Push images
-      if: env.need_to_build == 'true'
+      if: ${{ (env.need_to_build == 'true') && ((github.event_name != 'pull_request') || (github.repository == github.event.pull_request.head.repo.full_name)) }}
```
Comment (andrewsavage1) [inline .github/actions/docker_win/action.yaml:70]: Same as above

**Review by andrewsavage1 — CHANGES_REQUESTED** [review]

File: .github/actions/docker/action.yaml [hunk 5]
```diff
@@ -39,11 +40,23 @@ runs:
...
     # to use tag generated by the image build.
     - name: Set Docker Tag
-      if: ${{ (steps.changed-files.outputs.any_changed == 'false') && (github.event_name == 'pull_request') }}
+      id: set-docker-tag-presubmit-non-fork
+      if: ${{ (steps.changed-files.outputs.any_changed == 'false') && (github.event_name == 'pull_request') && (!github.event.pull_request.head.repo.fork) }}
       run: echo "DOCKER_TAG=ghcr.io/${{github.repository}}/${{inputs.docker_image}}:${GITHUB_BASE_REF%.1+}" >> $GITHUB_ENV
       shell: bash
+    - name: Set Docker Tag
+      id: set-docker-tag-presubmit-fork
+      if: ${{ (steps.changed-files.outputs.any_changed == 'true') (github.event_name == 'pull_request') && (github.event.pull_request.head.repo.fork) }}
```
Comment (andrewsavage1) [inline .github/actions/docker/action.yaml:49]: Missing an "and" or "or"

**Review by andrewsavage1 — CHANGES_REQUESTED** [review]

**Review by andrewsavage1 — APPROVED** [review]

**Pull request merged**